File: template/internal/swagger/member.go
// Copyright 2019 Brad Rydzewski. All rights reserved.
// Use of this source code is governed by the Polyform License
// that can be found in the LICENSE.md file.

package swagger

import (
	"net/http"

	"github.com/{{toLower repo}}/types"
)

// swagger:route GET /{{toLower project}}s/{{`{`}}{{toLower project}}{{`}`}}/members/{user} member getMember
//
// Get the {{toLower project}} member with the matching email address.
//
//     Responses:
//       200: member
//
func memberFind(w http.ResponseWriter, r *http.Request) {}

// swagger:route GET  /{{toLower project}}s/{{`{`}}{{toLower project}}{{`}`}}/members member getMemberList
//
// Get the list of all {{toLower project}} members.
//
//     Responses:
//       200: memberList
//
func memberList(w http.ResponseWriter, r *http.Request) {}

// swagger:route POST /{{toLower project}}s/{{`{`}}{{toLower project}}{{`}`}}/members member createMember
//
// Create a new {{toLower project}} member.
//
//     Responses:
//       200: member
//
func memberCreate(w http.ResponseWriter, r *http.Request) {}

// swagger:route PATCH /{{toLower project}}s/{{`{`}}{{toLower project}}{{`}`}}/members/{user} member updateMember
//
// Update the {{toLower project}} member.
//
//     Responses:
//       200: member
//
func memberUpdate(w http.ResponseWriter, r *http.Request) {}

// swagger:route DELETE /{{toLower project}}s/{{`{`}}{{toLower project}}{{`}`}}/members/{user} member deleteMember
//
// Delete the {{toLower project}} member.
//
//     Responses:
//       204:
//
func memberDelete(w http.ResponseWriter, r *http.Request) {}

// swagger:parameters getMember deleteMember
type memberReq struct {
	// in: path
	{{title project}} int64 `json:"{{toLower project}}"`

	// in: path
	Email string `json:"user"`
}

// swagger:parameters getMemberList
type memberListReq struct {
	// in: path
	{{title project}} int64 `json:"{{toLower project}}"`
}

// swagger:parameters createMember
type memberCreateReq struct {
	// in: path
	{{title project}} int64 `json:"{{toLower project}}"`

	// in: body
	Body types.MembershipInput
}

// swagger:parameters updateMember
type memberUpdateReq struct {
	// in: path
	{{title project}} int64 `json:"{{toLower project}}"`

	// in: path
	Email string `json:"user"`

	// in: body
	Body types.MembershipInput
}

// swagger:response member
type memberResp struct {
	// in: body
	Body types.Member
}

// swagger:response memberList
type memberListResp struct {
	// in: body
	Body []types.Member
}

File: template/web/public/api/member.js
import { instance } from "./config.js";
import useSWR, { mutate } from "swr";

/**
 * createMember creates a new member.
 */
export const createMember = ({{toLower project}}, data, fetcher) => {
	return fetcher(
		`${instance}/api/v1/{{toLower project}}s/${{`{`}}{{toLower project}}{{`}`}}/members/${data.email}`,
		{
			body: JSON.stringify(data),
			method: "POST",
		}
	).then((member) => {
		mutate(`${instance}/api/v1/{{toLower project}}s/${{`{`}}{{toLower project}}{{`}`}}/members`);
		return member;
	});
};

/**
 * updateMember updates an existing member.
 */
export const updateMember = ({{toLower project}}, member, data, fetcher) => {
	return fetcher(`${instance}/api/v1/{{toLower project}}s/${{`{`}}{{toLower project}}{{`}`}}/members/${member}`, {
		body: JSON.stringify(data),
		method: "PATCH",
	});
};

/**
 * deleteMember deletes an existing member.
 */
export const deleteMember = ({{toLower project}}, member, fetcher) => {
	return fetcher(`${instance}/api/v1/{{toLower project}}s/${{`{`}}{{toLower project}}{{`}`}}/members/${member}`, {
		method: "DELETE",
	}).then((response) => {
		mutate(`${instance}/api/v1/{{toLower project}}s/${{`{`}}{{toLower project}}{{`}`}}/members`);
		return response;
	});
};

/**
 * use returns an swr hook that provides
 */
export const useMemberList = ({{toLower project}}) => {
	const { data, error } = useSWR(
		`${instance}/api/v1/{{toLower project}}s/${{`{`}}{{toLower project}}{{`}`}}/members`
	);

	return {
		memberList: data,
		isLoading: !error && !data,
		isError: error,
	};
};
